src/lua/protobuf.lua

```diff
@@ -0,0 +1,723 @@
+    assert(type(data) == 'table')
+    local key_count = 0
+    local min_key = 1/0
+    local max_key = -1/0
+    for k, _ in pairs(data) do
+        if type(k) ~= 'number' then
+            error(('Input array for %q repeated field ' ..
+                'contains non-numeric key: %q'):format(field_def.name, k))
+        end
+        if k - math.floor(k) ~= 0 then
+            error(('Input array for %q repeated field contains ' ..
+                'non-integer numeric key: %q'):format(field_def.name, k))
+        end
+        key_count = key_count + 1
+        min_key = math.min(min_key, k)
+        max_key = math.max(max_key, k)
+    end
+    if key_count == 0 then
+        return
+    end
+    if min_key ~= 1 then
+        error(('Input array for %q repeated field got min index %d. ' ..
+            'Must be 1'):format(field_def.name, min_key))
+    end
+    if max_key ~= key_count then
+        error(('Input array for %q repeated field has inconsistent keys. ' ..
+            'Got table with %d fields and max index of %d'):format(
+            field_def.name, key_count, max_key))
+    end
+end
+
+local function validate_type(field_def, value, exp_type)
+    if type(exp_type) == 'table' then
+        local found = false
+        for _, exp_t in pairs(exp_type) do
+            if type(value) == exp_t then
+                found = true
+                break
+            end
+        end
+        if not found then
+            error(('Field %q of %q type gets %q type value.'):format(
+                field_def.name, field_def.type, type(value)))
+        end
+        return
+    end
+    assert(type(exp_type) == 'string')
+    if type(value) ~= exp_type then
+        error(('Field %q of %q type gets %q type value. ' ..
+            'Unsupported or colliding types'):format(field_def.name,
+            field_def.type, type(value)))
+    end
+    return
+end
+
+local function validate_value_error(field_def, value)
+    error(('Input data for %q field is %q and do not fit in %q')
+        :format(field_def.name, value, field_def.type))
+end
+
+local function validate_value(field_def, value, min, max)
+    -- To avoid incorrect comparison unsigned cdata is not
+    -- checked for lower limit
+    if type(value) == 'cdata' and ffi.istype(uint64_t, value) then
+        if value > max then
+            validate_value_error(field_def, value)
+        end
+        return
+    end
+    if value < min or value > max then
+        validate_value_error(field_def, value)
+    end
+end
+
+-- }}} Validations
+
+
+-- {{{ Scalar type definitions
+
+scalars.float = {
+    encode_as_packed = true,
+    validate = function(value, field_def)
+        local MAX_FLOAT = 3.4028234E+38
+        local MIN_FLOAT = -3.4028234E+38
+        validate_type(field_def, value, 'number')
+        validate_value(field_def, math.abs(value), MIN_FLOAT, MAX_FLOAT)
+    end,
+    encode = function(value, field_id)
+        return wireformat.encode_float(value, field_id)
+    end,
```

Comment:
In most of the cases we don't need an extra wrapping function.

<details>
<summary>Diff</summary>

```diff
diff --git a/src/lua/protobuf.lua b/src/lua/protobuf.lua
index c80eb537e5..8ce701a9b9 100644
--- a/src/lua/protobuf.lua
+++ b/src/lua/protobuf.lua
@@ -333,9 +333,7 @@ scalars.float = {
         validate_type(field_def, value, 'number')
         validate_value(field_def, math.abs(value), MIN_FLOAT, MAX_FLOAT)
     end,
-    encode = function(value, field_id)
-        return wireformat.encode_float(value, field_id)
-    end,
+    encode = wireformat.encode_float,
 }
 
 scalars.fixed32 = {
@@ -355,9 +353,7 @@ scalars.fixed32 = {
             validate_value(field_def, value, MIN_CDATA, MAX_CDATA)
         end
     end,
-    encode = function(value, field_id)
-        return wireformat.encode_fixed32(value, field_id)
-    end,
+    encode = wireformat.encode_fixed32
 }
 
 scalars.sfixed32 = {
@@ -377,9 +373,7 @@ scalars.sfixed32 = {
             validate_value(field_def, value, MIN_CDATA, MAX_CDATA)
         end
     end,
-    encode = function(value, field_id)
-        return wireformat.encode_fixed32(value, field_id)
-    end,
+    encode = wireformat.encode_fixed32,
 }
 
 scalars.double = {
@@ -390,9 +384,7 @@ scalars.double = {
         validate_type(field_def, value, 'number')
         validate_value(field_def, value, MIN_DOUBLE, MAX_DOUBLE)
     end,
-    encode = function(value, field_id)
-        return wireformat.encode_double(value, field_id)
-    end,
+    encode = wireformat.encode_double,
 }
 
 scalars.fixed64 = {
@@ -414,9 +406,7 @@ scalars.fixed64 = {
             validate_value(field_def, value, MIN_CDATA, MAX_CDATA)
         end
     end,
-    encode = function(value, field_id)
-        return wireformat.encode_fixed64(value, field_id)
-    end,
+    encode = wireformat.encode_fixed64,
 }
 
 scalars.sfixed64 = {
@@ -438,9 +428,7 @@ scalars.sfixed64 = {
             validate_value(field_def, value, MIN_CDATA, MAX_CDATA)
         end
     end,
-    encode = function(value, field_id)
-        return wireformat.encode_fixed64(value, field_id)
-    end,
+    encode = wireformat.encode_fixed64,
 }
 
 scalars.string = {
@@ -449,7 +437,7 @@ scalars.string = {
         validate_type(field_def, value, 'string')
         validate_length(value)
     end,
-    encode = wireformat.encode_len
+    encode = wireformat.encode_len,
 }
 
 scalars.bytes = scalars.string
@@ -500,9 +488,7 @@ scalars.uint32 = {
         check_integer(field_def, value)
         validate_value(field_def, value, MIN_UINT32, MAX_UINT32)
     end,
-    encode = function(value, field_id)
-        return wireformat.encode_uint(value, field_id)
-    end,
+    encode = wireformat.encode_uint,
 }
 
 scalars.int64 = {
@@ -583,9 +569,7 @@ scalars.uint64 = {
                'uint_64'):format(field_def.name, tonumber(value)))
         end
     end,
-    encode = function(value, field_id)
-        return wireformat.encode_uint(value, field_id)
-    end,
+    encode = wireformat.encode_uint,
 }
 
 scalars.bool = {
```
</details>

It is easier to read when it is just `encode = wireformat.encode_xxx`.

Reply:
`sint32` and `sint64` don't need it too and it fixed a bug in the encoding (see a comment to the wireformat module).

`int32` and `int64` don't need it too, at least because the `ffi.cast()` is inside wireformat's code.

<details>
<summary>Incremental diff</summary>

```diff
diff --git a/src/lua/protobuf.lua b/src/lua/protobuf.lua
index 684a076ac5..f32fe158f4 100644
--- a/src/lua/protobuf.lua
+++ b/src/lua/protobuf.lua
@@ -451,14 +451,7 @@ scalars.int32 = {
         check_integer(field_def, value)
         validate_value(field_def, value, MIN_INT32, MAX_INT32)
     end,
-    encode = function(value, field_id)
-        if value >= 0 then
-            return wireformat.encode_uint(value, field_id)
-        elseif value < 0 then
-            return wireformat.encode_uint(
-                ffi.cast('uint64_t', value), field_id)
-        end
-    end,
+    encode = wireformat.encode_uint,
 }
 
 scalars.sint32 = {
@@ -505,14 +498,7 @@ scalars.int64 = {
             validate_value(field_def, value, MIN_CDATA, MAX_CDATA)
         end
     end,
-    encode = function(value, field_id)
-        if value >= 0 then
-            return wireformat.encode_uint(value, field_id)
-        elseif value < 0 then
-            return wireformat.encode_uint(
-                ffi.cast('uint64_t', value), field_id)
-        end
-    end,
+    encode = wireformat.encode_uint,
 }
 
 scalars.sint64 = {
```
</details>

Reply:
Nice!

The bug that you mention here seems resolved by changing the `encode_varint` implementation. I'm closing the discussion so.